### ISSUE TITLE
Measuring tool

### ISSUE THREAD

**Issue by FedeClaudi** [brainglobe/brainreg-segment#42]
Proposal to add a new widget to brainreg segment which can be used to measure distances/angles between arbitrary points.

The way I envisioned it is to use a `points` layer to let the use select N points in the image and compute the length and angles (AP and ML) of each segment between two consecutive points. 
The results could be saved to file with the coordinates and brain region of each segment's end points as well as the measured quantities.

**Comment by juanma9613** [comment]
Hi, I'd like to work on this feature as well, do you still want help for this one?

**Comment by adamltyson** [comment]
Hi @juanma9613, yes definitely!

The existing `brainreg-segment` widget is already doing quite a lot, so it might be better to have a simple, dedicated widget for this, what do you think @FedeClaudi? The downside is we'd have to make sure that any other data that's needed (e.g. atlas images) are available outside of the main widget.

**Comment by juanma9613** [comment]
That's great! In the meantime I'm exploring the tutorial and reading a little bit about Brainreg. Please let me know if you have any guidance about the design of the widget before starting.

**Comment by adamltyson** [comment]
I don't have any specific guidance at the moment, but feel free to open a draft PR and I can provide feedback. If you need any test data etc. let me know.

Thanks for taking a look at this!

**Comment by FedeClaudi** [comment]
Hey, 
I see advantages for both having it as part of brainreg-segment and separate. I'm sure it's a tool that could be widely useful for anyone using napari, but at the same time within the scope of brainglobe it would realistically only be used to measure distances and angles between points in image/atlas space, so it would be handy to have it load the data like brainreg-segment.

I'm not familiar with plugins, can you have more than one 'active' at a given time? As in, can you use brainreg-segment to load the data and XXX to measure things? Or can you have a plugin use another plugin?

Finally, the way I was doing it while playing around with this idea was to use a `points` layer and accessing it's `data` attribute to get the points coordinates. You can then compute distance/angle for each consecutive pair of point. I guess ideally it'd generate an output in which you have for each pair of points their coordinates, brain region they are in, distance and angles.

Computing distance is trivial: `np.linalg.norm(p1-p0)`, but it might be harder to deal with angles. I think ideally you'd compute the angle on the AP and ML planes separately as that's what normally used e.g. to guide surgeries, but perhaps there's better ways?

**Comment by adamltyson** [comment]
>I'm not familiar with plugins, can you have more than one 'active' at a given time? 

Yep, you can have >1 widget in the same plugin, and these can be added to napari independently. 

>As in, can you use brainreg-segment to load the data and XXX to measure things? Or can you have a plugin use another plugin? 

To do this, I think we'd make sure that the new widget could just work on napari layers, which would be output by the original widget. We could add any arbitrary "extra" data or information needed as layer metadata.

I haven't put much thought into exactly what this plugin would calculate, but your ideas sound good @FedeClaudi

**Comment by juanma9613** [comment]
Hi, 

During the weekend I was working on the first piece of code for this tool. At the moment it's just a small script with a demo of usage and two functions to get distances between all the points and the angles of the segments joining two consecutive 
points. Docs and unit testing are still pending.

I've been reading the discussion but I don't know yet how to incorporate the functions I just created. Maybe I can start by creating it as a separate Napari plugin, [following this](https://napari.org/plugins/stable/for_plugin_developers.html#plugins-for-plugin-developers), and then we can figure out how to plug it either to Brainglobe as a widget or to brainreg-segment directly, What do you think about this idea? 

Although it's just preliminary work (mainly to get used to Napari inside python) please let me know if you have comments about the code below.

```python
import math
from typing import List, Tuple

import napari
import numpy as np
from skimage.data import astronaut
from scipy.spatial import distance_matrix

# Defining the two functions to use to get distances and angles
def get_distances_between_point_layers(point_layer: napari.layers.Points) -> np.array:
    return distance_matrix(point_layer.data, point_layer.data)

# For the axis I will assume the following (x,y,z) are each of the following
#  (anteroposterior, mediolateral, vectical)
def get_angles_between_consecutive_points(point_layer: napari.layers.Points) -> Tuple[List[float], List[float]]:
    # TODO: Assumed points are always 3d, be careful when that's not the case
    points_array  = point_layer.data
    n_points = points_array.shape[0]
    ap_angles = []
    mp_angles = []
    for i in range(1, n_points):
        pointA = points_array[i-i, :]
        pointB = points_array[i, :]
        segment_vector = pointB - pointA
        # projecting the vector in the sagital plane is just removing
        #  the "y/mediolateral" coordinate
        segment_vector_zx = segment_vector[[2,0]]
        # for the angle, z is used as the first direction/axis and x as the other 
        ap_angle = math.atan2(segment_vector_zx[1], segment_vector_zx[0])
        ap_angles.append(ap_angle)
        # projecting the vector in the frontal plane is just removing
        #  the anterioposterior coordinate.
        segment_vector_zy = segment_vector[[2,1]]
        # for the angle, z is used as the first direction/axis and y as the other 
        mp_angle = math.atan2(segment_vector_zy[1], segment_vector_zy[0])
        mp_angles.append(mp_angle)

    return ap_angles, mp_angles

# create a Viewer and add an image here
viewer = napari.view_image(astronaut(), rgb=True)
# add random points to the view
points = np.array([[100, 100, 0], [200, 200, 1], [300, 100, 50]])
points_layer = viewer.add_points(points, size=30)

points_distance_matrix = get_distances_between_point_layers(points_layer)
ap_angles, mp_angles = get_angles_between_consecutive_points(points_layer)

print('distances', points_distance_matrix)
print('angles', ap_angles, mp_angles)

# start the event loop and show the viewer
napari.run()
```

**Comment by adamltyson** [comment]
Hi @juanma9613,

From a quick look, that looks good!

Your idea is a good one. If you get this up and running as a napari widget, we can decide later whether it should live on it's own, or whether it should be part of brainreg-segment. It's straightforward for one plugin to have >1 widget.

I would have a play with [magicgui](https://napari.org/magicgui/) to create a GUI that allows the user to choose a points layer, and run the function, perhaps reporting the results in a separate window, or saving to disk. The widget could also optionally run on e.g. only selected points, or >1 points layer.

**Comment by juanma9613** [comment]
Unfortunately it was a busy month but this is my first PR for this tool.
Unit tests and some checks to make the plugin more robust to errors are pending, but I'd like to know if the functionality is approved. 
 https://github.com/brainglobe/brainreg-segment/pull/43
 
 thanks again.
